sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,519 @@
+            _maxLength = maxLength;
+            _cancellationToken = cancellationToken;
+        }
+
+        public async ValueTask LogAsync(string requestId, bool isError, Stream? stream, Encoding? textEncoding, bool async)
+        {
+            EventType eventType = ResponseOrError(isError);
+
+            if (stream == null || !IsEnabled(eventType))
+            {
+                return;
+            }
+
+            var bytes = await FormatAsync(stream, async).ConfigureAwait(false).EnsureCompleted(async);
+            Log(requestId, eventType, bytes, textEncoding);
+        }
+
+        public async ValueTask LogAsync(string requestId, BinaryContent? content, Encoding? textEncoding, bool async)
+        {
+            EventType eventType = EventType.Request;
+
+            if (content == null || !IsEnabled(eventType))
+            {
+                return;
+            }
+
+            var bytes = await FormatAsync(content, async).ConfigureAwait(false).EnsureCompleted(async);
+
+            Log(requestId, eventType, bytes, textEncoding);
+        }
+
+        public void Log(string requestId, bool isError, byte[] buffer, int offset, int length, Encoding? textEncoding, int? block = null)
+        {
+            EventType eventType = ResponseOrError(isError);
+
+            if (buffer == null || !IsEnabled(eventType))
+            {
+                return;
+            }
+
+            var logLength = Math.Min(length, _maxLength);
+
+            byte[] bytes;
+            if (length == logLength && offset == 0)
+            {
+                bytes = buffer;
+            }
+            else
+            {
+                bytes = new byte[logLength];
+                Array.Copy(buffer, offset, bytes, 0, logLength);
+            }
+
+            Log(requestId, eventType, bytes, textEncoding, block);
+        }
+
+        public bool IsEnabled(bool isError)
+        {
+            return IsEnabled(ResponseOrError(isError));
+        }
+
+        private bool IsEnabled(EventType errorResponse)
+        {
+            return _eventSource != null &&
+                   (_eventSource.IsEnabled(EventLevel.Informational, EventKeywords.All) ||
+                   (errorResponse == EventType.ErrorResponse && _eventSource.IsEnabled(EventLevel.Warning, EventKeywords.All)));
+        }
+
+        private async ValueTask<byte[]> FormatAsync(BinaryContent requestContent, bool async)
+        {
+            using var memoryStream = new MaxLengthStream(_maxLength);
+
+            if (async)
+            {
+                await requestContent.WriteToAsync(memoryStream, _cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                requestContent.WriteTo(memoryStream, _cancellationToken);
+            }
+
+            return memoryStream.ToArray();
+        }
+
+        private async ValueTask<byte[]> FormatAsync(Stream content, bool async)
+        {
+            content.Seek(0, SeekOrigin.Begin);
```

Comment:
Curious: it looks like this was added in the ClientModel type, but not present in the Azure.Core implementation.  What makes it needed in ClientModel and not Azure.Core?